### PR TITLE
subrange `get<0>` requires a copyable iterator

### LINKED ISSUE #2399: LWG-3589 The `const` lvalue reference overload of `get` for `subrange` does not constrain `I` to be `copyable` when `N == 0`
LWG-3589 The `const` lvalue reference overload of `get` for `subrange` does not constrain `I` to be `copyable` when `N == 0`

This was voted into the Working Paper in October 2021, see #2236.

### PR DESCRIPTION
**Pull request by CaseyCarter**
per LWG-3589.

Fixes #2399

### CODE REVIEW

**Review by miscco — APPROVED** [review]
I sah what you did there

File: stl/inc/xutility [hunk 1]
```diff
@@ -3487,7 +3487,7 @@ namespace ranges {
 
     // clang-format off
     template <size_t _Idx, class _It, class _Se, subrange_kind _Ki>
-        requires (_Idx < 2)
+        requires ((_Idx  == 0 && copyable<_It>) || _Idx == 1)
```
Comment (StephanTLavavej) [inline stl/inc/xutility:3490]: There's an extra space here. (The peril of disabling clang-format! :scream_cat:)

File: stl/inc/xutility [hunk 2]
```diff
@@ -3497,7 +3497,7 @@ namespace ranges {
     }
 
     template <size_t _Idx, class _It, class _Se, subrange_kind _Ki>
-        requires (_Idx < 2)
+        requires ((_Idx  == 0 && copyable<_It>) || _Idx == 1)
     _NODISCARD constexpr auto get(subrange<_It, _Se, _Ki>&& _Val) {
```
Comment (StephanTLavavej) [inline stl/inc/xutility:3501]: `get(subrange&&)` should not be changed according to LWG-3589.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/xutility (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/xutility]: There's "I'm in a hurry trying to implement all the Ranges LWG issues" and then there's "I saw two patches of green so I changed two things without concern for what two things they happened to be".

File: tests/std/tests/P0896R4_ranges_subrange/test.compile.pass.cpp [hunk 3]
```diff
@@ -1483,3 +1484,44 @@ namespace test_subrange {
     STATIC_ASSERT(test_tuple<subrange<int*, std::unreachable_sentinel_t, subrange_kind::sized>>());
     STATIC_ASSERT(test_tuple<subrange<int*, std::unreachable_sentinel_t, subrange_kind::unsized>>());
 } // namespace test_subrange
+
+namespace test_lwg_3589 {
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_subrange/test.compile.pass.cpp:1488]: "LWG" is capitalized on 26.

Comment (StephanTLavavej) [reply tests/std/tests/P0896R4_ranges_subrange/test.compile.pass.cpp:1488]: I can live with this, no need to reset testing.

**Review by StephanTLavavej — APPROVED** [review]

**Review by mnatsuhara — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this LWG issue resolution! :white_check_mark: :rocket: :tada:

**Pull request merged**